daemon/containerio/cio.go

```diff
@@ -2,229 +2,191 @@ package containerio
 
 import (
 	"context"
-	"fmt"
 	"io"
 	"io/ioutil"
 	"os"
 	"path/filepath"
 	"sync"
 	"syscall"
 
-	containerdio "github.com/containerd/containerd/cio"
+	"github.com/containerd/containerd/cio"
 	"github.com/containerd/fifo"
+	"github.com/sirupsen/logrus"
 )
 
-// cio is a basic container IO implementation.
-type cio struct {
-	config containerdio.Config
-
-	closer *wgCloser
-}
-
-func (c *cio) Config() containerdio.Config {
-	return c.config
-}
-
-func (c *cio) Cancel() {
-	if c.closer == nil {
-		return
-	}
-	c.closer.Cancel()
-}
-
-func (c *cio) Wait() {
-	if c.closer == nil {
-		return
-	}
-	c.closer.Wait()
+// CioFIFOSet holds the fifo pipe for containerd shim.
+//
+// It will be replaced by the containerd@v1.2
+type CioFIFOSet struct {
+	cio.Config
+	closeFn func() error
 }
 
-func (c *cio) Close() error {
-	if c.closer == nil {
-		return nil
+// Close will remove all the relatived files.
+func (c *CioFIFOSet) Close() error {
+	if c.closeFn != nil {
+		return c.closeFn()
 	}
-	return c.closer.Close()
+	return nil
 }
 
-// NewFifos returns a new set of fifos for the task
-func NewFifos(id string, stdin bool) (*containerdio.FIFOSet, error) {
+// NewCioFIFOSet prepares fifo files.
+func NewCioFIFOSet(processID string, withStdin bool, withTerminal bool) (*CioFIFOSet, error) {
 	root := "/run/containerd/fifo"
 	if err := os.MkdirAll(root, 0700); err != nil {
 		return nil, err
 	}
-	dir, err := ioutil.TempDir(root, "")
+
+	fifoDir, err := ioutil.TempDir(root, "")
 	if err != nil {
 		return nil, err
 	}
-	fifos := &containerdio.FIFOSet{
-		Dir: dir,
-		In:  filepath.Join(dir, id+"-stdin"),
-		Out: filepath.Join(dir, id+"-stdout"),
-		Err: filepath.Join(dir, id+"-stderr"),
+
+	cfg := cio.Config{
+		Terminal: withTerminal,
+		Stdout:   filepath.Join(fifoDir, processID+"-stdout"),
+	}
+
+	if withStdin {
+		cfg.Stdin = filepath.Join(fifoDir, processID+"-stdin")
 	}
 
-	if !stdin {
-		fifos.In = ""
+	if !withTerminal {
+		cfg.Stderr = filepath.Join(fifoDir, processID+"-stderr")
```

Comment:
I do not think stderr fifo has related with terminal is open

Reply:
Yes. I checked that containerd-shim. The shim will open socket as pty-slave. There is only one output from pty-slave so that we don't need stderr fifo if the `withTerminal` is true